sdks/java/core/src/main/java/org/apache/beam/sdk/io/WholeFileIO.java

```diff
@@ -0,0 +1,347 @@
+    }
+
+    @Override
+    public PDone expand(PCollection<KV<String, byte[]>> input) {
+      checkNotNull(
+          getOutputDir(),
+          "Need to set the output directory of a WholeFileIO.Write transform."
+      );
+
+      input.apply(
+          ParDo.of(
+              new DoFn<KV<String, byte[]>, Void>() {
+                ValueProvider<ResourceId> tmpDir;
+
+                @Setup
+                public void setup() {
+                  tmpDir = ValueProvider.NestedValueProvider.of(
+                      getOutputDir(),
+                      new TemporaryDirectoryBuilder()
+                  );
+                }
+
+                @ProcessElement
+                public void processElement(ProcessContext c) throws IOException {
+                  KV<String, byte[]> kv = c.element();
+
+                  String filename = kv.getKey();
+                  ResourceId tmpFile = tmpDir.get().resolve(filename, RESOLVE_FILE);
+
+                  byte[] bytes = kv.getValue();
+                  try (
+                      OutputStream outStream =
+                          Channels.newOutputStream(FileSystems.create(tmpFile, BINARY))
+                  ) {
+                    outStream.write(bytes);
+                    outStream.flush();
+                  } catch (IOException e) {
+                    LOG.error(
+                        "Failed to write to temporary file [{}] for [{}].",
+                        tmpFile,
+                        getOutputDir().get().resolve(filename, RESOLVE_FILE)
+                    );
+                    FileSystems.delete(
+                        Collections.singletonList(tmpFile),
+                        MoveOptions.StandardMoveOptions.IGNORE_MISSING_FILES
+                    );
+                    throw e;
+                  }
+                }
+
+                @Teardown
+                public void teardown() throws IOException {
+                  try {
+                    // TODO: Replace FileSystems.rename() with a copy and delete to eliminate
+                    // empty directory restriction on Write.
+                    FileSystems.rename(
+                        Collections.singletonList(tmpDir.get()),
+                        Collections.singletonList(getOutputDir().get())
+                    );
+                  } catch (IOException e) {
+                    LOG.error(
+                        "Failed to rename temporary directory [{}] to [{}].",
+                        tmpDir.get(), getOutputDir().get()
+                    );
+                    throw e;
+                  }
+                }
+              }
+          )
+      );
+
+      return PDone.in(input.getPipeline());
+    }
+
+    // TODO: Either fully customize this tool for WholeFileIO or change
+    // FileBasedSink.WriteOperation.TemporaryDirectoryBuilder from private to public (or protected).
+    private static class TemporaryDirectoryBuilder
```

Comment:
This is copied almost exactly from FileBasedSink.WriteOperation, because it's marked private in FileBasedSink.

We could change the FileBasedSink implementation to public or protected, or I could specialize this logic for my purposes and keep the pseudo-duplicate in WholeFileIO.